src/mqtt/protocol/mqtt/mqtt_client.c

```diff
@@ -0,0 +1,1110 @@
+					scheduler = true;
+				} else {
+					// scheduling disorder, PUBACK received
+					work_set_final(work);
+				}
+			} else {
+				// for QoS 2, expect receiving a PUBREC
+				if (!work_is_acked(work)) {
+					work_set_recv(work, NNG_MQTT_PUBREC);
+					scheduler = true;
+				} else {
+					// scheduling disorder, PUBREC received
+					work_set_recv(work, NNG_MQTT_PUBCOMP);
+				}
+			}
+		}
+		break;
+
+	case NNG_MQTT_PUBREC:
+		// we have sent a PUBREC, expect receiving a PUBREL
+		work_set_recv(work, NNG_MQTT_PUBREL);
+		break;
+
+	case NNG_MQTT_PUBREL:
+		// we have sent a PUBREL, expect receiving a PUBCOMP
+		work_set_recv(work, NNG_MQTT_PUBCOMP);
+		break;
+
+	case NNG_MQTT_SUBSCRIBE:
+		// we have sent a SUBSCRIBE, expect receiving a SUBACK
+		if (!work_is_acked(work)) {
+			work_set_recv(work, NNG_MQTT_SUBACK);
+			scheduler = true;
+		} else {
+			// scheduling disorder, SUBACK received
+			work_set_final(work);
+		}
+		break;
+
+	case NNG_MQTT_UNSUBSCRIBE:
+		// we have sent a UNSUBSCRIBE, expect receiving a UNSUBACK
+		if (!work_is_acked(work)) {
+			work_set_recv(work, NNG_MQTT_UNSUBACK);
+			scheduler = true;
+		} else {
+			// scheduling disorder, UNSUBACK received
+			work_set_final(work);
+		}
+		break;
+
+	default:
+		work_set_error(work);
+		break;
+	}
+
+	if (work_is_error(work)) {
+		// MQTT protocol error, terminate the connection
+		nni_aio_finish_error(work->user_aio, NNG_EPROTO);
+		nni_mtx_unlock(&s->mtx);
+		nni_pipe_close(p->pipe);
+		return;
+	} else if (work_is_final(work)) {
+		// good news, protocol state machine run to the end
+		nni_aio *aio = work->user_aio;
+		mqtt_sock_free_work(s, work);
+		mqtt_send_start(s);
+		nni_mtx_unlock(&s->mtx);
+		if (NULL != aio) {
+			nni_aio_finish(aio, 0, 0);
+		}
+		return;
+	}
+
+	mqtt_send_start(s);
+	nni_mtx_unlock(&s->mtx);
+	if (scheduler) {
+		work_timer_schedule(work);
+	}
+	return;
+}
+
+static void
+mqtt_recv_cb(void *arg)
+{
+	mqtt_pipe_t *p = arg;
+	mqtt_sock_t *s = p->mqtt_sock;
+	work_t *     work;
+
+	nni_mtx_lock(&s->mtx);
+
+	nni_msg *msg = nni_aio_get_msg(&p->recv_aio);
+
+	if (nni_aio_result(&p->recv_aio) != 0) {
+		nni_msg_free(msg);
+		nni_mtx_unlock(&s->mtx);
+		nni_pipe_close(p->pipe);
+		return;
+	}
+
+	if (nni_atomic_get_bool(&s->closed) ||
+	    nni_atomic_get_bool(&p->closed)) {
+		nni_mtx_unlock(&s->mtx);
+		return;
+	}
+
+	nni_aio_set_msg(&p->recv_aio, NULL);
+	nni_msg_set_pipe(msg, nni_pipe_id(p->pipe));
+	nni_mqtt_msg_proto_data_alloc(msg);
+	nni_mqtt_msg_decode(msg);
+
+	packet_type_t packet_type = nni_mqtt_msg_get_packet_type(msg);
+	int32_t       packet_id;
+	uint8_t       qos;
+
+	// schedule another receive
+	nni_pipe_recv(p->pipe, &p->recv_aio);
+
+	// state transitions
+	switch (packet_type) {
+	case NNG_MQTT_CONNACK:
+		// we have received the CONNACK
+		nni_mtx_unlock(&s->mtx);
+		return;
+
+	case NNG_MQTT_PUBACK:
+		// we have received a PUBACK, successful delivery of a QoS 1
+
+		// fall through
```

Comment:
I think we should say "FALLTHROUGH" -- I expect some checkers are looking for that specific text.

Reply:
done